timelock-agent/src/main/java/com/palantir/timelock/paxos/AsyncTimeLockServicesCreator.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.paxos;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.timelock.AsyncTimelockResource;
+import com.palantir.atlasdb.timelock.AsyncTimelockService;
+import com.palantir.atlasdb.timelock.AsyncTimelockServiceImpl;
+import com.palantir.atlasdb.timelock.TimeLockServices;
+import com.palantir.atlasdb.timelock.config.AsyncLockConfiguration;
+import com.palantir.atlasdb.timelock.lock.AsyncLockService;
+import com.palantir.atlasdb.timelock.lock.NonTransactionalLockService;
+import com.palantir.atlasdb.timelock.paxos.ManagedTimestampService;
+import com.palantir.atlasdb.timelock.util.AsyncOrLegacyTimelockService;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.atlasdb.util.JavaSuppliers;
+import com.palantir.lock.RemoteLockService;
+
+public class AsyncTimeLockServicesCreator implements TimeLockServicesCreator {
+    private final PaxosLeadershipCreator leadershipCreator;
+    private final AsyncLockConfiguration asyncLockConfiguration;
+
+    public AsyncTimeLockServicesCreator(PaxosLeadershipCreator leadershipCreator,
+            AsyncLockConfiguration asyncLockConfiguration) {
+        this.leadershipCreator = leadershipCreator;
+        this.asyncLockConfiguration = asyncLockConfiguration;
+    }
+
+    @Override
+    public TimeLockServices createTimeLockServices(
+            String client,
+            Supplier<ManagedTimestampService> rawTimestampServiceSupplier,
+            Supplier<RemoteLockService> rawLockServiceSupplier) {
+        AsyncOrLegacyTimelockService asyncOrLegacyTimelockService;
+        AsyncTimelockService asyncTimelockService = instrumentInLeadershipProxy(
+                AsyncTimelockService.class,
+                () -> AsyncTimeLockServicesCreator.createRawAsyncTimelockService(client, rawTimestampServiceSupplier),
+                client);
+        asyncOrLegacyTimelockService = AsyncOrLegacyTimelockService.createFromAsyncTimelock(
+                new AsyncTimelockResource(asyncTimelockService));
+
+        Supplier<RemoteLockService> lockServiceSupplier =
+                asyncLockConfiguration.disableLegacySafetyChecksWarningPotentialDataCorruption()
+                        ? rawLockServiceSupplier
+                        : JavaSuppliers.compose(NonTransactionalLockService::new, rawLockServiceSupplier);
+
+        return TimeLockServices.create(
+                asyncTimelockService,
+                instrumentInLeadershipProxy(RemoteLockService.class, lockServiceSupplier, client),
```

Comment:
nit: can we move the leader proxy wrapping above, so the code structure for wrapping is more uniform for lock / timestamp?